### PR TITLE
Adding Feature Ablation

### PR DESCRIPTION
**Pull request by vivekmig**
This adds feature ablation, which replaces each feature with a baseline value. Occlusion is a small variant of this, which covers a fixed region surrounding a particular pixel, and will be added in a separate PR.

### CODE REVIEW

File: captum/attr/_core/feature_ablation.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+            )
+            yield tuple(current_features), current_additional_args, current_mask
+            # Replace existing tensor at index i.
+            current_features[i] = original_tensor
+            num_features_processed += current_batch
+
+    def _construct_ablated_input(
+        self, feature_tensor, input_mask, baseline, start_feature, end_feature
+    ):
+        r"""
```
Comment (vivekmig) [inline captum/attr/_core/feature_ablation.py]: @miguelmartin75 I think you should be able to implement shuffling on top of this by just overriding this function to shuffle each batch instead of replacing with the baseline.

File: captum/attr/_core/feature_ablation.py [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+                            - a tuple of scalars, if inputs is a tuple of tensors, with
+                                exactly the same number of elements as inputs tuple.
+                                Each scalar element in baselines' tuple is broadcasted
+                                for each input tensor at the same index in inputs
+                                tuple.
+                            Default: zero scalar for each input tensor
+                internal_batch_size (int, optional): Divides total #features *
+                            #examples data points into chunks of size
+                            internal_batch_size, which are computed sequentially.
+                            internal_batch_size should be a multiple of #examples.
```
Comment (miguelmartin75) [inline captum/attr/_core/feature_ablation.py]: Is this condition being enforced?

Comment (vivekmig) [reply captum/attr/_core/feature_ablation.py]: No, it's just implicitly being rounded down to the nearest multiple of #examples. But that's a good point, will add a line of documentation for this, thanks!

File: captum/attr/_core/feature_ablation.py [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+
+                            - a tuple of scalars, if inputs is a tuple of tensors, with
+                                exactly the same number of elements as inputs tuple.
+                                Each scalar element in baselines' tuple is broadcasted
+                                for each input tensor at the same index in inputs
+                                tuple.
+                            Default: zero scalar for each input tensor
+                internal_batch_size (int, optional): Divides total #features *
+                            #examples data points into chunks of size
+                            internal_batch_size, which are computed sequentially.
```
Comment (miguelmartin75) [inline captum/attr/_core/feature_ablation.py]: I think the documentation on this parameter should be slightly modified to be more clear, such that it mentions:

- this parameter is the maximum number of examples that can be fed into the feed-forward network 
- if `internal_batch_size > num_examples`, the algorithm will compute multiple ablations simultaneously? e.g. if `internal_batch_size = 2*num_examples` then 2 features will be ablated/evaluated for each feed forward.

File: captum/attr/_core/feature_ablation.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+                            the given argument is used for all forward evaluations.
+                            Note that attributions are not computed with respect
+                            to these arguments.
+                            Default: None
+                feature_mask (tensor or tuple of tensors, optional):
+                            feature_mask defines a mask for the input, grouping
+                            features which should be ablated together. feature_mask
+                            should contain the same number of tensors as inputs,
+                            and features within each input tensor are ablated
+                            independetly (not across tensors). Each tensor should
```
Comment (miguelmartin75) [inline captum/attr/_core/feature_ablation.py]: independetly -> independently

File: captum/attr/_core/feature_ablation.py [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+                ),
+                input_mask,
+                baseline,
+                num_features_processed,
+                num_features_processed + current_batch,
+            )
+            current_features[i] = ablated_features.reshape(
+                (-1,) + ablated_features.shape[2:]
+            )
+            yield tuple(current_features), current_additional_args, current_mask
```
Comment (NarineK) [inline captum/attr/_core/feature_ablation.py]: Can we maybe add more description what the dimensionality of `current_features` is ? Does this represent a #batches x #features  ... ? 
Also, that `num_features_processed` is being updated by `current_batch `, maybe it is the naming ,because by current_batch I imagine that it contains all features for given batch.

File: captum/attr/_core/feature_ablation.py [hunk 6]
```diff
@@ -0,0 +1,329 @@
+#!/usr/bin/env python3
+
+import torch
+
+from .._utils.common import (
+    _format_attributions,
+    format_input,
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/feature_ablation.py]: Sure, will change once it merges!

File: captum/attr/_core/feature_ablation.py [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+        super().__init__(forward_func)
+
+    def attribute(
+        self,
+        inputs,
+        target=None,
+        additional_forward_args=None,
+        feature_mask=None,
+        baselines=None,
+        internal_batch_size=None,
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/feature_ablation.py]: Good point, will fix!

File: captum/attr/_core/feature_ablation.py [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+                            - A single tuple, which contains #output_dims - 1
+                                elements. This target index is applied to all examples.
+
+                            - A list of tuples with length equal to the number of
+                                examples in inputs (dim 0), and each tuple containing
+                                #output_dims - 1 elements. Each tuple is applied as the
+                                target for the corresponding example.
+
+                            Default: None
+                additional_forward_args (tuple, optional): If the forward function
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/feature_ablation.py]: Good point, yeah this has to be fixed everywhere. For consistency, I'll leave it this way for now, and create a separate PR that fixes it everywhere once this merges?

File: captum/attr/_core/feature_ablation.py [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+        be broadcastable to match feature_tensor.
+
+        This method returns the ablated feature tensor, which has the same
+        dimensionality as feature_tensor as well as the corresponding mask with
+        either the same dimensionality as feature_tensor or second dimension
+        being 1. This mask contains 1s in locations which have been ablated (and
+        thus counted towards ablations for that feature).
+        """
+        current_mask = torch.stack(
+            [input_mask == j for j in range(start_feature, end_feature)], dim=0
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/feature_ablation.py]: This will inherit device from input_mask, but will make sure input_mask device always matches inputs, thanks!

File: captum/attr/_core/feature_ablation.py [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+    _format_attributions,
+    format_input,
+    _run_forward,
+    _expand_additional_forward_args,
+    _format_additional_forward_args,
+)
+from .._utils.attribution import PerturbationAttribution
+
+
+class FeatureAblation(PerturbationAttribution):
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/feature_ablation.py]: Yes, good point! Since I'm also adding other perturbation based algos, I was going to add a section for these in the readme and website, will add them together in a separate PR if that's ok?

File: captum/attr/_core/feature_ablation.py [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+            >>> # | 0 | 0 | 1 | 1 |
+            >>> # +---+---+---+---+
+            >>> # | 0 | 0 | 1 | 1 |
+            >>> # +---+---+---+---+
+            >>> # | 2 | 2 | 3 | 3 |
+            >>> # +---+---+---+---+
+            >>> # | 2 | 2 | 3 | 3 |
+            >>> # +---+---+---+---+
+            >>> # With this mask, all inputs with the same value are ablated
+            >>> # simultaneously, and the attribution for all inputs in the same
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/feature_ablation.py]: Good point, will update the wording to make it clearer, thanks!
